pkg/ffmpeg/stream.go

```diff
@@ -0,0 +1,273 @@
+}
+
+type TranscodeStreamOptions struct {
+	ProbeResult      VideoFile
+	Codec            Codec
+	Hls              bool
+	StartTime        string
+	MaxTranscodeSize models.StreamingResolutionEnum
+	// transcode the video, remove the audio
+	// in some videos where the audio codec is not supported by ffmpeg
+	// ffmpeg fails if you try to transcode the audio
+	VideoOnly bool
+}
+
+func GetTranscodeStreamOptions(probeResult VideoFile, videoCodec string, audioCodec AudioCodec, container Container, supportedVideoCodecs []string) TranscodeStreamOptions {
+	options := TranscodeStreamOptions{
+		ProbeResult: probeResult,
+	}
+
+	options.setTranscodeCodec(supportedVideoCodecs)
+
+	if audioCodec == MissingUnsupported {
+		//ffmpeg fails if it trys to transcode a non supported audio codec
+		options.VideoOnly = true
+	} else {
+		// try to be smart if the video to be transcoded is in a Matroska container
+		// mp4 has always supported audio so it doesn't need to be checked
+		// while mpeg_ts has seeking issues if we don't reencode the video
+
+		// If MKV is supported and video codec is also supported then only transcode audio
+		if IsValidCodec(Mkv, supportedVideoCodecs) && Container(container) == Matroska && IsValidCodec(videoCodec, supportedVideoCodecs) {
+			// copy video stream instead of transcoding it
+			options.Codec = CodecMKVAudio
+		}
+	}
+
+	return options
+}
+
+func (o *TranscodeStreamOptions) setTranscodeCodec(supportedVideoCodecs []string) {
+	if len(supportedVideoCodecs) == 0 {
+		supportedVideoCodecs = DefaultSupportedCodecs
+	}
+
+	logger.Debugf("Choosing transcode codec from: %s", strings.Join(supportedVideoCodecs, ","))
+
+	// TODO - make preferred order configurable
+	if IsValidCodec(Vp9, supportedVideoCodecs) {
+		logger.Debug("Using VP9")
+		o.Codec = CodecVP9
+	} else if IsValidCodec(Vp8, supportedVideoCodecs) {
+		logger.Debug("Using VP8")
+		o.Codec = CodecVP8
+	} else if IsValidCodec(Hevc, supportedVideoCodecs) {
+		logger.Debug("Using HEVC")
+		o.Codec = CodecHEVC
+	} else if IsValidCodec(Hls, supportedVideoCodecs) {
+		logger.Debug("Using HLS (with H264)")
+		o.Codec = CodecHLS
+		o.Hls = true
+	} else {
+		logger.Debug("Using H264")
+		o.Codec = CodecH264
+	}
+}
+
+func (o TranscodeStreamOptions) getStreamArgs() []string {
+	scale := calculateTranscodeScale(o.ProbeResult, o.MaxTranscodeSize)
+
+	args := []string{
+		"-hide_banner",
+		"-v", "error",
+	}
+
+	if o.StartTime != "" {
+		args = append(args, "-ss", o.StartTime)
+	}
+
+	if o.Hls {
+		// we only serve a fixed segment length
+		args = append(args, "-t", strconv.Itoa(int(hlsSegmentLength)))
+	}
+
+	args = append(args,
+		"-i", o.ProbeResult.Path,
+	)
+
+	if o.VideoOnly {
+		args = append(args, "-an")
+	}
+
+	args = append(args,
+		"-c:v", o.Codec.Codec,
+		"-vf", "scale="+scale,
+	)
+
+	if len(o.Codec.extraArgs) > 0 {
+		args = append(args, o.Codec.extraArgs...)
+	}
+
+	args = append(args,
+		// this is needed for 5-channel ac3 files
+		"-ac", "2",
+		"-b:v", "0",
```

Comment:
`"-b:v", "0",` this should be removed from there and added as an extraArg in the VP8 and VP9 profile.
It is needed there of a quirk in libvpx 
It shouldn't be used in other settings because it might interfere with the crf setting